sklearn/model_selection/split.py

```diff
@@ -0,0 +1,1237 @@
+#         Olivier Girsel <olivier.grisel@ensta.org>
+# License: BSD 3 clause
+
+
+from __future__ import print_function
+from __future__ import division
+
+import warnings
+import inspect
+from itertools import chain, combinations
+from collections import Iterable
+from math import ceil, floor
+import numbers
+from abc import ABCMeta, abstractmethod
+
+import numpy as np
+
+from scipy.misc import comb
+from ..utils import indexable, check_random_state, safe_indexing
+from ..utils.validation import _num_samples, column_or_1d
+from ..utils.multiclass import type_of_target
+from ..externals.six import with_metaclass
+from ..externals.six.moves import zip
+from ..utils.fixes import bincount
+from ..base import _pprint
+
+__all__ = ['BaseCrossValidator',
+           'KFold',
+           'LeaveOneLabelOut',
+           'LeaveOneOut',
+           'LeavePLabelOut',
+           'LeavePOut',
+           'ShuffleSplit',
+           'StratifiedKFold',
+           'StratifiedShuffleSplit',
+           'PredefinedSplit',
+           'train_test_split',
+           'check_cv']
+
+
+class BaseCrossValidator(with_metaclass(ABCMeta)):
+    """Base class for all cross-validators
+
+    Implementations must define `_iter_test_masks` or `_iter_test_indices`.
+    """
+
+    def split(self, X, y=None, labels=None):
+        """Generate indices to split data into training and test set.
+
+        Parameters
+        ----------
+        X : array-like, shape (n_samples, n_features)
+            Training data, where n_samples is the number of samples
+            and n_features is the number of features.
+
+        y : array-like, shape (n_samples,)
+            The target variable for supervised learning problems.
+
+        labels : array-like of int with shape (n_samples,), optional
+            Arbitrary domain-specific stratification of the data to be used
+            to draw the splits.
+        """
+        X, y, labels = indexable(X, y, labels)
+        ind = np.arange(_num_samples(X))
+        for test_index in self._iter_test_masks(X, y, labels):
+            train_index = ind[np.logical_not(test_index)]
+            test_index = ind[test_index]
+            yield train_index, test_index
+
+    # Since subclasses must implement either _iter_test_masks or
+    # _iter_test_indices, neither can be abstract.
+    def _iter_test_masks(self, X=None, y=None, labels=None):
+        """Generates boolean masks corresponding to test sets.
+
+        By default, delegates to _iter_test_indices(X, y, labels)
+        """
+        for test_index in self._iter_test_indices(X, y, labels):
+            test_mask = np.zeros(_num_samples(X), dtype=np.bool)
+            test_mask[test_index] = True
+            yield test_mask
+
+    def _iter_test_indices(self, X=None, y=None, labels=None):
+        """Generates integer indices corresponding to test sets."""
+        raise NotImplementedError
+
+    @abstractmethod
+    def n_splits(self, X=None, y=None, labels=None):
```

Comment:
On 08/10/2015 04:40 AM, Arnaud Joly wrote:
> > +    def n_splits(self, X=None, y=None, labels=None):
>
> This looks like what would be return when you call |len|.
>
But there can't be ``len`` because it depends on X and y, right?